README.md

```diff
@@ -1 +1,262 @@
+to produce the output.  Spark improved upon this by trying to keep everything
+in memory and doing more intelligent inter-cluster shuffles to avoid round-trips
+to disk if and when possible.
+
+Both hadoop and spark implemented the fully general map-reduce model.
+But what if we focused on zql-style queries leveraging the typical case of
+composable aggregations and integrate into the design the use of search indexes
+during query traversals?  And what if this were all exposed as simple
+unix tools that can be experimented with from the command-line without
+setting up complex clusters?
+
+### version 1: a static model
+
+The hadoopy approach for all this is to just spill to disk.  Any aggregation
+(or sort etc)
+in a zql flowgraph will have a table, organized by a set of group-by or
+sort keys (one or more of primary, second, etc key).  To deal with
+memory limits, whenever this table hits a configured size limit (related
+to the amount of memory dedicated to the worker),
+it is simply spilled to disk by writing the entries to disk
+in a key-sorted order.  At end-of-stream, the tables are all opened
+and read in parallel and any entry that is common across the streams is
+aggregated by the aggregation operator and the results are streamed
+to the downstream entity.
+
+This approach can be parallelized across N nodes where the time range of the
+traversal is divided amongst the different nodes and the final step merges
+the streamed results from all of the nodes.
+
+That will work fine for many use cases.
+For many aggregations, the group-by table will stay small when the cardinality
+of the key space is low.
+If you have a feel for how much memory you need ahead of time, you could
+launch N jobs in parallel so a high-cardinality group-by table
+gets split out efficiently.  In this case, the key space isn't disjoint
+but it might not matter much if the sub-cardinilaty of each key is small
+(meaning there is little advantage to a shuffle)
+so in general this could work well depending on the underlying data
+characteristic.  And if the memory limit is hit here, the table is spilled
+to disk as above.
+
+### version 2: a dynamic model
+
+There are two key problems with the static approach.  First, you don't generally
+know how to choose N so it's unreasonable to ask the user to specify this.
+And second, the key space is replicated over each node making the tables inefficient
+and larger than they need to be.
+
+A solution to these problems is, firstly, to fork a new worker only after the
+aggregation table gets too big (thus providing scale-out memory along with
+scale-out throughput) and, secondly, to implement a shuffle to
+route the table rows to the worker that owns the row.  Ownership of rows
+is handled like spark and map-reduce using a hash partition.  The
+number of dynamics nodes would be limited either because the cluster
+manager (i.e., kubernetes) indicates no more resources are available
```

Comment:
```suggestion
manager indicates no more resources are available
```

Reply:
I'm going to change i.e. to e.g. and leave this to imply that we don't need a custom cluster manager like ray and spark have.